### ISSUE TITLE
Commands Generator for Model Based Testing

### ISSUE THREAD

**Issue by maciejmyslinski** [dubzzz/fast-check#250]
I recently played with Model-Based Testing. I noticed that if a precondition fails, the framework tries to apply the same command with different arguments. This is unfortunate since this might be the case that a particular command cannot be applied as the first one.

Current behavior: if a command cannot be applied, try to generate new arguments for the same command.
Expected behavior: if a command cannot be applied, try to generate entirely new command (including new arugments)

**Comment by dubzzz** [comment]
Would be interested in having more details on the commands you wanted to generate and the problem. A minimal failing case would be very helpful to understand more precisely the problem

**Comment by maciejmyslinski** [comment]
I wanted to test a dynamic form.

**Initial state**
There are no inputs.
There is a button to add an input.
There is a button to move inputs (like drag and drop reordering).

**Commands**
* add input
* change value
* move input

As you can see, at the beginning only one command can be executed (no matter the arguments) – add input. We cannot execute change value or move input as the first command since there are no inputs to work with.

**Comment by dubzzz** [comment]
In order to illustrate my previous message, here is a recent example of commands applied to Selenium UI tests: https://github.com/dubzzz/connect-four-react/blob/master/e2e/commands

Here are the preconditions for those commands:

- CheckEndOfGameCommand: user cannot play anymore and grid is not full
- CheckPlayerTurnCommand: user can play somewhere
- NewGameCommand: always possible
- PlayForbiddenTokenCommand(columnIdx): columnIdx is not playable
- PlayTokenCommand(columnIdx): columnIdx is playable

**Comment by dubzzz** [comment]
Did I answer your question?

I personally use `fc.commands` at work to test a SPA, a pro tip is to increase the number of generated commands - `fc.commands([...], upperBoundForNumberGenCommands)` - so that you increase your chances to encounter a bug (default is 10, as for `fc.array`).

**Comment by maciejmyslinski** [comment]
The link https://github.com/dubzzz/connect-four-react/blob/master/e2e/commands is not working.
Thank you for the explanation!
So you say if the `check` condition fails, fast-check will disregard the command and try to generate a new one?
How does it work? Does it first generate all commands and then run them one by one (check first, then if check returns true, the command)? Or does it generate, run, and then generate the next command, and run, and so on?

**Comment by dubzzz** [comment]
Just updated the link with the correct link: https://github.com/dubzzz/connect-four-react/tree/master/e2e/limited-model/commands

**Model based testing in fast-check**

In order to benefit from the built-in model-based testing tools provided by fast-check you should use: `fc.commands` in conjonction with `fc.modelRun` (or its async equivalent).

When called on its `generate` method (by `fc.assert`), `fc.commands` produces a `Shrinkable` (value + how to shrink it) of kind `Iterable<Command<Model, Real>>`. In reality today, the returned instance is an `Array<Command<Model, Real>>`.

The array contains multiple commands, among those commands some might be invalid at some times and should be disgarded if so.

In a nutshell, `fc.modelRun` can be summarized as follow:

```js
function modelRun(setup, commands) {
  const { model, real } = setup();
  for (const cmd of commands) {
    if (cmd.check(model)) {
      cmd.run(model, real);
    }
  }
}
```

Where `run` is supposed to throw on error (`expect` or `assert` do throw by default in modern test frameworks - except Jasmine).

Then in case of failure, fast-check has to shrink the array of commands. In order to be a bit more clever, the shrink takes into account the list of commands that were really executed, the others are just removed because they are not relevant. Then it shrinks.

**Comment by dubzzz** [comment]
@maciejmyslinski any updates?

**Comment by maciejmyslinski** [comment]
I make a lot of sense, thank you @dubzzz for clarification and patience! 🙇 👍 💯
